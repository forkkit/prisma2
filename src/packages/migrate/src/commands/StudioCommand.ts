import { arg, Command, format, HelpError, isError } from '@prisma/sdk'
import chalk from 'chalk'

import { Studio } from '../Studio'
import { ProviderAliases } from '@prisma/sdk'
import { ExperimentalFlagError } from '../utils/experimental'

export class StudioCommand implements Command {
  public static new(providerAliases: ProviderAliases): StudioCommand {
    return new StudioCommand(providerAliases)
  }

  private static help = format(`
    Browse your data with Studio

    ${chalk.bold.yellow('WARNING')} ${chalk.bold(
    "Prisma's studio functionality is currently in an experimental state.",
  )}
    ${chalk.dim(
      'When using any of the commands below you need to explicitly opt-in via the --experimental flag.',
    )}

    ${chalk.bold('Usage')}

      ${chalk.dim('$')} prisma studio --experimental

    ${chalk.bold('Options')}

      -h, --help        Displays this help message
      -p, --port        Port to start Studio on
      -b, --browser     Browser to open Studio in

    ${chalk.bold('Examples')}

      Start Studio on the default port
      ${chalk.dim('$')} prisma studio --experimental

      Start Studio on a custom port
      ${chalk.dim('$')} prisma studio --port 5555 --experimental

      Start Studio in a specific browser
      ${chalk.dim(
        '$',
      )} prisma studio --port 5555 --browser firefox --experimental
      ${chalk.dim('$')} BROWSER=firefox prisma studio --port 5555 --experimental

      Start Studio without opening in a browser
      ${chalk.dim('$')} prisma studio --port 5555 --browser none --experimental
      ${chalk.dim('$')} BROWSER=none prisma studio --port 5555 --experimental
  `)

  private constructor(private readonly providerAliases: ProviderAliases) {
    this.providerAliases = providerAliases
  }

  /**
   * Parses arguments passed to this command
   *
   * @param argv Array of all arguments
   */
  public async parse(argv: string[]): Promise<string | Error> {
    const args = arg(argv, {
      '--help': Boolean,
      '-h': '--help',
      '--port': Number,
      '-p': '--port',
      '--browser': String,
      '-b': '--browser',
      '--schema': String,
      '--experimental': Boolean,
    })

    if (isError(args)) {
      return this.help(args.message)
    }

    if (args['--help']) {
      return this.help()
    }

    if (!args['--experimental']) {
      throw new ExperimentalFlagError()
    }

    const port = args['--port'] || 5555
    const browser = args['--browser'] || process.env.BROWSER

    const studio = new Studio({
      schemaPath: args['--schema'],
      browser,
      port,
    })

    return studio.start(this.providerAliases)
  }

  // help message
  public help(error?: string): string | HelpError {
    if (error) {
      return new HelpError(
        `\n${chalk.bold.red(`!`)} ${error}\n${StudioCommand.help}`,
      )
    }

    return StudioCommand.help
  }
}
