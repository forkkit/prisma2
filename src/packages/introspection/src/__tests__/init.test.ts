import { join } from 'path'
import { promisify } from 'util'
import fs from 'fs'
import rimraf from 'rimraf'
import { Init, defaultSchema, defaultEnv } from '../commands/Init'
import stripAnsi from 'strip-ansi'

const del = promisify(rimraf)
const tmp = join(__dirname, '../../prisma')

beforeEach(async () => {
  await del(tmp)
})

afterAll(async () => {
  await del(tmp)
})

test('is schema and env written on disk replace', async () => {
  const init = Init.new()
  const result = stripAnsi(await init.parse([]))

  const schema = fs.readFileSync(join(tmp, 'schema.prisma'), 'utf-8')
  const env = fs.readFileSync(join(tmp, '.env'), 'utf-8')

  expect(schema).toMatch(defaultSchema)
  expect(env).toMatch(defaultEnv)
  expect(result).toMatchSnapshot(result)
})
