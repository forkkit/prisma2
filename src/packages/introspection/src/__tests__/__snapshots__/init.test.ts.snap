// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is schema and env written on disk replace: 
✔ Your Prisma schema was created at prisma/schema.prisma.
  You can now open it in your favorite editor.

Next steps:
1. Set the provider of the datasource block in schema.prisma to match your database: postgresql, mysql or sqlite.
2. Set the DATABASE_URL in the .env file to point to your existing database. If your database has no tables yet, read https://pris.ly/d/getting-started.
3. Run prisma introspect to turn your database schema into a Prisma data model.
4. Run prisma generate to install Prisma Client. You can then start querying your database.

More information in our documentation:
https://pris.ly/d/getting-started
     1`] = `
"
✔ Your Prisma schema was created at prisma/schema.prisma.
  You can now open it in your favorite editor.

Next steps:
1. Set the provider of the datasource block in schema.prisma to match your database: postgresql, mysql or sqlite.
2. Set the DATABASE_URL in the .env file to point to your existing database. If your database has no tables yet, read https://pris.ly/d/getting-started.
3. Run prisma introspect to turn your database schema into a Prisma data model.
4. Run prisma generate to install Prisma Client. You can then start querying your database.

More information in our documentation:
https://pris.ly/d/getting-started
    "
`;
